WIP: Editorial changes (#259)
tlswg/draft-ietf-tls-esni: draft-ietf-tls-esni.md

# Introduction

> [[TODO: Add a high-level description to this section of the security goal of

Reviewer: Agreed. Let's make an issue for this (and remove the TODO)!

Author: Done. https://github.com/tlswg/draft-ietf-tls-esni/issues/260

defines ({{encrypted-client-hello}}). Finally, it sends ClientHelloOuter to the
server.

Upon receiving the ClientHelloOuter, the client-facing server takes one of the
following actions:

1. If it does not support ECH, it ignores the "encrypted_client_hello" extension
   and proceeds with the handshake as usual, per {{RFC8446}}, Section 4.1.2.
1. If it supports ECH but cannot decrypt it, then it ignores the extension and
   proceeds with the handshake as usual. This is referred to as "ECH rejection".
   When ECH is rejected, the server sends an acceptable ECH configuration in its
   EncryptedExtensions message.
1. If it supports ECH and can decrypt it, then it forwards the CllientHelloInner
   to the backend, who terminates the connection. This is referred to as "ECH
   acceptance".

> [[TODO: Ensure this is specified below: What does the server do if it can

Reviewer: This is already described in the section you reference, so let's remove this TODO.

Author: Ack.

decrypt but decryption fails, i.e., the ciphertext is invalid? Does it abort or
proceed with the outer CH?]]

Upon receiving the server's response, the client determines whether ECH was
accepted or rejected and proceeds with the handshake accordingly. (See

> {{client-behavior}}.)

Reviewer: ```suggestion
{{client-behavior}} for details.)
```

Author: Oops, missed this. Shall I make this change/

defines ({{encrypted-client-hello}}). Finally, it sends ClientHelloOuter to the
server.

Upon receiving the ClientHelloOuter, the server takes one of the following
actions:

1. If the client-facing server does not support ECH, it ignores the
   "encrypted_client_hello" extension and proceeds with the handshake as usual,
   per {{RFC8446}}, Section 4.1.2.
1. If the client-facing server supports ECH but cannot decrypt it, then it
   ignores the extension and proceeds with the handshake as usual. This is
   referred to as "ECH rejection". When ECH is rejected, the server sends an
   acceptable ECH configuration in its EncryptedExtensions message.
1. If the client-facing server supports ECH and can decrypt it, then it forwards
   the CllientHelloInner to the backend, who terminates the connection. This is
   referred to as "ECH acceptance".

> Upon receiving the server's response, the client must first determine whether

Reviewer: I'm not sure this level of detail is needed here. Perhaps we can simply say, "The client determines whether ECH was accepted or rejected and proceeds accordingly. The details for each procedure are described in {{section}}."

Author: The high-level description is incomplete, since it doesn't talk about how the client processes the server's response. I think it's good to be complete, even if high-level.

Reviewer: What I proposed was high level, I think. We should avoid redundancies in the spec where possible.

Author: I think of it as a gentle introduction to the protocol flow, which, IMO, aids understanding later on. I think the slight redundancy is worth it. If you don't think it is, then I'll revert this.

Reviewer: Let's flush it out first and then we can decide if it's too deep? I commented on WIP text.

Author: Sounds good, I don't see your comment though.

Reviewer: It's the first comment in this thread :)

Author: Applied your suggested change.

the ClientHelloOuter or ClientHelloInner was used. [[TODO: Describe how this is
done. Currently this is by trial decryption.]] Once it makes this determination,
the client proceeds with the handshake, assuming the indicated ClientHello was
used by the backend server.

> [[TODO: Give a high-level description of "ech_nonce".]]

Reviewer: I don't think this is needed (here).

Author: As written, the "ech_nonce" section comes out of nowhere, since it's purpose isn't articulated until "security considerations". My view is that any code point defined in the spec needs to be mentioned in the overview. In fact, I would include "outer_extensions" here, too.

Reviewer: Sure, but this isn't the right place for it. It's a tiny part of the protocol. We can describe inline below when it arises.

Author: Ack, moving TODO to section on ech_nonce.

    ECHConfig ECHConfigs<1..2^16-1>;
~~~~

The `ECHConfigs` structure contains one or more `ECHConfig` structures in
decreasing order of preference. This allows a server to support multiple
versions of ECH and multiple sets of ECH parameters.

The `ECHConfig` structure contains the following fields:

version
: The version of ECH for which this configuration is used. Versioning is
described below in {{versioning}}. Clients MUST ignore any `ECHConfig` structure
with a version they do not support.

length
: The length, in bytes, of the next field.

contents
: An opaque byte string whose contents depend on the version of the structure.
For this specification, the contents are an `ECHConfigContents` structure.

> [[TODO: `contents` isn't a named field in the `ECHConfig` structure. I'm

Reviewer: That's true. We can rewrite this as a paragraph, e.g., saying "The remaining contents of the `ECHConfig` structure are an opaque byte string..."

Author: Hmmm, I'm curious though: shouldn't this field have a label?
https://tools.ietf.org/html/rfc8446#section-3.8

Reviewer: It's a union, basically, whose value depends on the type. I don't know what would be the desired label, so I'd be fine with the current text or what's proposed above.

Author: What I don't like about your proposal is that the `length` field gives the length of a nameless field. I think `content` would be a reasonable label.

Reviewer: I see little difference between that and `contents`, so, let's just keep the current text?

Author: Oh, "contents" is fine. The suggested change is to give the variant a label.

Author: Hmm, I'm not sure what to here. I'm resolving this and leaving the text as is.

assuming `contents` is meant to be the label of the variant, but that's not what
the spec says.]]

The `ECHConfigContents` structure contains the following fields:

public_name
: The non-empty name of client-facing server, i.e., the entity trusted to update
these encryption keys. This is used to repair misconfigurations, as described in
{{handle-server-response}}.

public_key
: The HPKE public key used by the client to encrypt ClientHelloInner.

kem_id
: The HPKE KEM identifier corresponding to `public_key`. Clients MUST ignore any
`ECHConfig` structure with a key using a KEM they do not support.

cipher_suites
: The list of HPKE AEAD and KDF identifier pairs clients can use for encrypting
ClientHelloInner.

maximum_name_length
: The largest name the server expects to support, if known. If this value is not

> known it can be set to zero, in which case clients SHOULD use the padding scheme

Reviewer: I think we should revert this change. The padding scheme specifically applies to the ClientHelloInner.

Author: Reverted.

names can be added or removed from the anonymity set during the lifetime of a
particular resource record value.

extensions
: A list of extensions that the client must take into consideration when
generating a ClientHello message. These are described below
({{config-extensions}}).

## Configuration Extensions {#config-extensions}

ECH configuration extensions are used to to provide room for additional
functionality as needed. See {{config-extensions-guidance}} for guidance on
which types of extensions are appropriate for this structure.

The format is as defined in {{RFC8446}}, Section 4.2.
The same interpretation rules apply: extensions MAY appear in any order, but
there MUST NOT be more than one extension of the same type in the extensions
block. An extension can be tagged as mandatory by using an extension type
codepoint with the high order bit set to 1. A client that receives a mandatory
extension they do not understand MUST reject the `ECHConfig` content.

Clients MUST parse the extension list and check for unsupported mandatory
extensions. If an unsupported mandatory extension is present, clients MUST
ignore the `ECHConfig`.

## Versions {#versioning}

> [[TODO: Replace this with an enumerated type? This is what's required for

Reviewer: I don't think this is necessary. The version field speaks for itself, and this section doesn't add much. Can we remove this?

Author: Yes, I'll revert this. However, it seems technically incorrect to me: using RFC 8446 as the authoritative standard, the `select` block is only defined for enumerated types. This might be splitting hairs though ... I'm not sure how strict RFCs are about this.

Author: Reverted.

    } ClientEncryptedCH;
~~~~

cipher_suite
: The cipher suite used to encrypt ClientHelloInner. This MUST match a value
provided in the corresponding `ECHConfig.cipher_suites` list.

config_digest
: A cryptographic hash of the `ECHConfig` structure from which the ECH key was
obtained, i.e., from the first byte of "version" to the end of the structure.
This hash is computed using the hash function associated with `cipher_suite`,
i.e., the corresponding HPKE KDF algorithm hash.

enc
: The HPKE encapsulated key, used by servers to decrypt the corresponding
`encrypted_ch` field.

encrypted_ch
: The serialized and encrypted ClientHelloInner structure, encrypted using HPKE
as described in {{send-ech}}.

When the client-facing server accepts ECH, it does not send this extension. When
it rejects, it adds an "encrypted_client_hello" extension to EncryptedExtensions
with the following structure as the payload:

~~~
    struct {
       ECHConfigs retry_configs;
    } ServerEncryptedCH;
~~~

retry_configs
: An ECHConfigs structure containing one or more ECHConfig structures, in
decreasing order of preference, to be used by the client in subsequent
connection attempts.

> This document also defines the "ech_required" alert, which clients MUST send

Author: This is meant to be clarifying, not a spec change. The "ech_required" alert MUST be sent in this case, correct?

Reviewer: 👍 yep!

~~~~

suite
: The cipher suite used to encrypt ClientHelloInner. This MUST match a value

> provided in the corresponding `ECHConfig.cipher_suites` list.  [[TODO: Rename

Reviewer: That's a good idea, yep!

Author: Done.

"suite" to "cipher_suite" to align with `ECHConfig`?]]

record_digest
: A cryptographic hash of the `ECHConfig` structure from which the ECH key was
obtained, i.e., from the first byte of "version" to the end of the structure.
This hash is computed using the hash function associated with `suite`, i.e., the

> corresponding HPKE KDF algorithm hash. [[TODO: "record_digest" is a misnomer,

Reviewer: How about `config_digest`?

Author: I like it. Changed.

since the value is not the hash of a record. What about something more
descriptive, like "ech_config_id"?]]

enc
: The HPKE encapsulated key, used by servers to decrypt the corresponding

> encrypted_ch field. [[TODO: A more descriptive name would be helpful. What about

Reviewer: `enc` matches HPKE, so I think we should keep it as is.

Author: Ack, removing TODO

"encapsulated_key"?]]

encrypted_ch
: The serialized and encrypted ClientHelloInner structure, encrypted using HPKE
as described in {{send-ech}}.

> When the client-facing server accepts ECH, it does not send a response. When it

Reviewer: ```suggestion
When the client-facing server accepts ECH, it does not send this extension. When it
```

~~~

retry_configs
: An ECHConfigs structure containing one or more ECHConfig structures, in
decreasing order of preference, to be used by the client in subsequent
connection attempts.

> This document also defines the "ech_required" alert, which is sent by the client

Reviewer: ```suggestion
This document also defines the "ech_required" alert, which clients MUST send
```

when it offered an "encrypted_client_hello" extension that was not accepted by
the server.
[[TODO: Is this a "MAY", "SHOULD", or "MUST"?]]

> [[TODO: Add a reference to a point in the document that provides more context.]]

Reviewer: 👍

Author: Done.

# The "ech_nonce" extension {#ech-nonce}

When using ECH, the client MUST include an extension of type "ech_nonce" to the
ClientHelloInner (but not to the ClientHelloOuter). This nonce is used to ensure
that the server's encrypted Certificate can only be read by the entity which

> sent this ClientHello. [[TODO: It's not clear from context what this means.]]

Reviewer: Can we remove this TODO and make it an issue?

Author: This is addressed by https://github.com/tlswg/draft-ietf-tls-esni/issues/253. But I think it might be worth resolving this issue as part of this PR. Or should we wait?

Reviewer: Let's wait.

Author: Ack, removing TODO.

~~~~

nonce
: A 16-byte nonce exported from the HPKE encryption context, as described in
{{send-ech}}.

Per {{client-behavior}} and {{server-behavior}}, implementations are required to
track, alongside each PSK established by a previous connection, whether the
connection negotiated this extension with the "ech_accept" response type. If so,
this is referred to as an "ECH PSK". Otherwise, it is a "non-ECH PSK". This may
be implemented by adding a new field to client and server session states.

> [[TODO: Why does discussion of the PSK belong in this section, which is about

Reviewer: It probably doesn't. Can you find a better home for it?

Author: Not until I understand the PSK story. I'll leave this unresolved for now.

Reviewer: Let's remove the TODO then.

Author: Done.

"ech_nonce"?]]

> [[TODO: The "'ech_accept' response type" hasn't been defined here.]]

Reviewer: Good point. We should probably shuffle things around.

Reviewer: Let's either remove the TODO and file an issue, or address this here.

Author: Actually, this is the *only* place in the document where "ech_accept" appears. This needs to be defined.

Author: Is this an issue, or shall we define the alert here? I'm not sure how to do this.

Reviewer: Let's just make it an issue. It'll be a small change.

Author: Done. https://github.com/tlswg/draft-ietf-tls-esni/issues/267

ClientHelloOuter. The server then computes a hash of the reconstructed
ClientHelloInner. If the hash does not equal OuterExtensions.hash, the server
aborts the connection with an "illegal_parameter" alert.

[[TODO: What does the hash accomplish? It seems superfluous to me.]]

Clients SHOULD only use this mechanism for extensions which are large. All other
extensions SHOULD appear in both ClientHelloInner and ClientHelloOuter even if
they have identical values. [[TODO: Why the second SHOULD? If I were
implementing this, I would shallow-copy all redundant extensions, except maybe
those that have 0-length payloads.]]

# Client Behavior {#client-behavior}

## Sending an encrypted ClientHello {#send-ech}

In order to send an encrypted ClientHello, the client first determines if it
supports the server's chosen KEM, as identified by `ECHConfig.kem_id`. If one is
supported, the client MUST select an appropriate `HpkeCipherSuite` from the list
of suites offered by the server. If the client does not support the
corresponding KEM or is unable to select an appropriate group or
`HpkeCipherSuite`, it SHOULD ignore that `ECHConfig` value and MAY attempt to

> use another value provided by the server. [[TODO: "appropriate group": Isn't

Reviewer: I'm not sure I understand the question here. I don't think the text suggests that the group is part of the ciphersuite. We can probably s/"an appropriate group or"/"a"

Author: That works for me.
Done.

supports the server's chosen KEM, as identified by `ECHConfig.kem_id`. If one is
supported, the client MUST select an appropriate `HpkeCipherSuite` from the list
of suites offered by the server. If the client does not support the
corresponding KEM or is unable to select a suitable `HpkeCipherSuite`, it SHOULD
ignore that `ECHConfig` value and MAY attempt to use another value provided by
the server. The client MUST NOT send ECH using HPKE algorithms not advertised by
the server.

Given a compatible `ECHConfig` with fields `public_key` and `kem_id`, carrying
the `HpkePublicKey` and KEM identifier corresponding to the server, clients
compute an HPKE encryption context as follows:

~~~
    pkR = HPKE.KEM.Unmarshal(ECHConfig.public_key)
    enc, context = SetupBaseS(pkR, "tls13-ech")
    ech_nonce_value = context.Export("tls13-ech-nonce", 16)
    ech_hrr_key = context.Export("tls13-ech-hrr-key", 16)
~~~

Note that the HPKE algorithm identifiers are those which match the client's
chosen preference from `ECHConfig.cipher_suites`. The client MAY replace any
large, duplicated extensions in ClientHelloInner with the corresponding

> "outer_extensions" extension, as described in {{outer-extensions}}. [[TODO: This

Reviewer: That seems reasonable -- thanks!

Author: Done.

When sending ClientHello, the client first computes ClientHelloInner, including
any PSK binders, and then MAY substitute extensions which it knows will be
duplicated in ClientHelloOuter. To do so, the client computes a hash of the
entire ClientHelloInner message with the same hash as for the KDF used to

> encrypt ClientHelloInner. [[TODO: Will there always be a hash function

Reviewer: Yes -- each KDF has an associated hash function.

Author: From the spec, it seems like all that's required of the KDF is that it's an "extract-then-expand" KDF:
https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hpke-05#section-4

In fact, the currently assigned code points are all for HKDF:
https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hpke-05#section-7.2

But AFAICT there's no requirement that the KDF always have some underlying hash function. As an alternative to HKDF, one can imagine a dedicated primitive that satisfies the extract-then-expand API but is not constructed from a hash function.

Sorry if this is nit-picky :)

Reviewer: I see your point. Let's replace this with `Extract` instead of `Hash` and call it a day.

Author: Well, that would change the protocol. Better to make it an issue?

Reviewer: Yep!

Author: Done. https://github.com/tlswg/draft-ietf-tls-esni/issues/261

associated with the KDF? I.e., does HPKE use HKDF?]] Then, the client removes
and replaces extensions from ClientHelloInner with a single "outer_extensions"
extension. The list of `outer_extensions` include those which were removed from
ClientHelloInner, in the order in which they were removed. The hash contains the
full ClientHelloInner hash computed above.

This process is reversed by client-facing server. Specifically,
the server replaces the `outer_extensions` with extensions contained in
ClientHelloOuter. The server then computes a hash of the reconstructed
ClientHelloInner. If the hash does not equal OuterExtensions.hash, the server
aborts the connection with an "illegal_parameter" alert.

> [[TODO: This procedure destroys the intended order of the extensions. Say the

Reviewer: I'm not sure this is correct -- let's remove this and track it in a separate issue.

Author: Done. https://github.com/tlswg/draft-ietf-tls-esni/issues/262

extensions in the outer CH are [a, b, c, d] and I want to shallow-copy b in the
inner CH. What I would send is I send [a, b, c, encrypted_client_hello = [a, c,
d, outer_extensions = [ shallow_copy_of_b ]]]. The client-facing server doesn't
know the order in which the client intended them to be delivered.]]

> [[TODO: What does the hash accomplish? It seems superfluous to me.]]

Reviewer: It makes sure that the outer extension contents are not modified.

Author: Is there a specific attack this prevents?

Reviewer: I don't think so. We didn't have analyze this case (yet), so the safe thing was just hash to prevent changes. On reflection, any change in the outer extension values would change the transcript if the inner CH was selected, so maybe it's not needed? This is worth an issue.

Author: Done. https://github.com/tlswg/draft-ietf-tls-esni/issues/263

Clients SHOULD only use this mechanism for extensions which are large. All other
extensions SHOULD appear in both ClientHelloInner and ClientHelloOuter even if
they have identical values. [[TODO: Why the second SHOULD? If I were

> implementing this, I would shallow-copy all redundant extensions, except maybe

Reviewer: By redundant do you mean identical?

Author: I mean that I would shallow-copy any extension that appears in both the inner and outer CH. The computational overhead is low, it reduces the size of the outer CH, and it seems simpler than shallow-copying select things. (Suppose someone else invents another big extension after ECH is implemented.)

Reviewer: Right, identical. This would be a change, so let's move it to an issue.

Author: Done. https://github.com/tlswg/draft-ietf-tls-esni/issues/265.

"encrypted_client_hello" extension with the values as indicated above. In
particular,

- cipher_suite contains the client's chosen HpkeCipherSuite;
- config_digest contains the digest of the corresponding ECHConfig structure;
- enc contains the encapsulated key as output by SetupBaseS; and
- encrypted_ch contains the HPKE encapsulated key (enc) and the ClientHelloInner
  ciphertext (encrypted_ch_inner).

> [[TODO: This parameter list has already been defined. I suggest cutting it

Reviewer: This specifies how the parameters are populated, which is different from their definitions above. Let's keep it for now.

Author: Ack, removing TODO.

from here.]]

The client MUST place the value of `ECHConfig.public_name` in the

> ClientHelloOuter "server_name" extension. [[TODO: What does this mean? "The

Reviewer: It means the outer CH SNI value contains "ECHConfig.public_name".

Author: Ack, resolving.

client MUST place the value of `ECHConfig.public_name` with the value of
'server_name' extension"?]] The ClientHelloOuter MUST NOT contain a
"cached_info" extension {{!RFC7924}} with a CachedObject entry whose
CachedInformationType is "cert", since this indication would divulge the true
server name.

> [[TODO: It seems like this should be worked out more in detail in security

Reviewer: That seems reasonable. Let's file an issue to expand on those details, and remove this TODO.

Author: Done. https://github.com/tlswg/draft-ietf-tls-esni/issues/266

considerations. There might be other (future) extensions that interact with ECH
in a way this document can't control.]]
The remaining contents of the ClientHelloOuter MAY be identical to those in
ClientHelloInner but MAY also differ.

## Recommended Padding Scheme {#padding}

> [[TODO: I don't understand the purpose of the padding scheme. Is this meant to

Reviewer: It's a RECOMMENDED padding scheme. What's the confusion?

Author: I'll resolve this TOOD for now.

if a client proposes ALPN values in ClientHelloInner, the server-selected value
will be returned in an EncryptedExtension, so that handshake message also needs
to be padded using TLS record layer padding.

> [[TODO: I'm thrown by "will also need". This seems like a required behavior

Reviewer: Again, this is a RECOMMENDED policy. Clients could alternatively pad their inner CH to 1024B, or something. Padding is needed to hide size metadata. It doesn't affect the protocol mechanics.

Author: Resolving for now.
